pkg/cache/application.go

```diff
@@ -507,6 +514,19 @@ func (app *Application) handleCompleteApplicationEvent(event *fsm.Event) {
 	// TODO app lifecycle updates
 }
 
+func (app *Application) handleFailApplicationEvent(event *fsm.Event) {
+	// unallocated task states include New, Pending and Scheduling
+	unalloc := app.GetNewTasks()
+	unalloc = append(unalloc, app.GetPendingTasks()...)
+	unalloc = append(unalloc, app.GetSchedulingTask()...)
```

Comment:
You have a deadlock here, because the application is already locked: https://github.com/apache/incubator-yunikorn-k8shim/blob/master/pkg/cache/application.go#L153-L155

Reply:
You can simply use `unalloc := app.getTasks(events.States().Task.New)` ...

Reply:
Thanks. Done.